src/olmo_core/train/callbacks/batch_size_scheduler.py

```diff
@@ -0,0 +1,171 @@
+import dataclasses
+import logging
+import math
+from dataclasses import dataclass
+from typing import List, Optional
+
+import torch
+
+from olmo_core.exceptions import OLMoConfigurationError
+from olmo_core.optim import INITIAL_LR_FIELD, LR_FIELD, Scheduler, SkipStepAdamW
+
+from ..common import Duration
+from ..train_module import TransformerPipelineTrainModule, TransformerTrainModule
+from .callback import Callback
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class BatchSizeSchedulerCallback(Callback):
+    """
+    A callback for setting a batch size scheduler over the course of a training run.
+    Also adjusts the base learning rate with Adam optimizers for transformer train modules by a factor of
+    ``sqrt(new_batch_size / current_batch_size)``.
+    """
+
+    batch_sizes: List[int] = dataclasses.field(default_factory=list)
+    """
+    Defines the batch sizes to apply, in order.
+    """
+
+    schedule: List[Duration] = dataclasses.field(default_factory=list)
+    """
+    Defines the schedule at which to apply each batch size.
+    """
+
+    def __post_init__(self):
+        if len(self.batch_sizes) != len(self.schedule):
+            raise OLMoConfigurationError(
+                "batch_sizes and schedules should have the same number of items"
+            )
+
+        if not self.schedule:
+            return
+
+        if len(set([duration.unit for duration in self.schedule])) > 1:
```

Comment:
Python has set comprehensions too. `{x for x in something}`.

Reply:
done: ff895e8f